src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -684,14 +683,42 @@ private void HandleSemanticTokens<T>(string id, T semanticTokensParams, string m
 
             var tokens = result.GetTokens(tokenTypesToSkip);
 
-            if (isRangeSemanticTokensMethod)
+            if (!isFullDocument)
             {
                 // Only consider overlapping tokens. end index is exlcusive
                 tokens = tokens.Where(token => !(token.EndIndex <= startIndex || token.StartIndex >= endIndex));
             }
 
-            var encodedTokens = SemanticTokensEncoder.EncodeTokens(tokens, expression, eol);
+            var controlTokenDict = isFullDocument ? new ControlTokenDictionary() : null;
+
+            var encodedTokens = SemanticTokensEncoder.EncodeTokens(tokens, expression, eol, controlTokenDict);
             _sendToClient(JsonRpcHelper.CreateSuccessResult(id, new SemanticTokensResponse() { Data = encodedTokens }));
+
+            PublishControlTokenNotification(controlTokenDict, queryParams);
+        }
+
+        /// <summary>
+        /// Handles publishing a control token notification if any control tokens found.
+        /// </summary>
+        /// <param name="controlTokenDict">Collection to add control tokens to.</param>
+        /// <param name="queryParams">Collection of query params.</param>
+        private void PublishControlTokenNotification(ControlTokenDictionary controlTokenDict, NameValueCollection queryParams)
+        {
+            if (controlTokenDict == null || controlTokenDict.Count() == 0 || queryParams == null)
```

Comment:
Low hanging fruit. But Count() would actually iterate to get the count. Alternatively you can just implement Count inside ControlTokenDictionary which just returns the underlying dictionary's size which doesn't require iterating